wire up everything

dknopik opened the pull request:
## Proposed Changes

- add config for network selection
- read or generate SSV private key
- init database and sync
- fill validator store from database

## Additional Info

:tada: :tada: :tada:

There is some room for optimisation/alignment between the components, but we can leave that for later.

dknopik committed:
wire up everything

dknopik committed:
Merge branch 'unstable' into wire-up

dknopik committed:
minor db fixes

- consistently use the same `OperatorId` if not known
- really only store own clusters in `clusters`

Zacholme7 commented on anchor/client/src/config.rs:
```diff
@@ -2,30 +2,31 @@
 
 pub const DEFAULT_BEACON_NODE: &str = "http://localhost:5052/";
 pub const DEFAULT_EXECUTION_NODE: &str = "http://localhost:8545/";
+pub const DEFAULT_EXECUTION_NODE_WS: &str = "wss://localhost:8545/";
```
For default local nodes, I believe this should just be "ws"

Zacholme7 commented on anchor/client/src/lib.rs:
```diff
@@ -276,26 +295,70 @@ impl Client {
+        );
 
-        // Create the processor-adjacent managers
+        // Wait until we have an operator id
```
We need to make sure we are in live sync first before getting the operator id. For example, if we are operator 1 we will get our id right away and we will continue executing while we are still syncing.

The easiest way to do this is to probably work in an `AtomicBool` that is flipped when we enter live sync. Once the bool is true we know we are synced up and then can either fetch or wait on the id.

dknopik replied:
Ah - realizing now that SSV supports removing a validator from the SSV network without exiting it :sweat_smile:. Yeah, then waiting to be fully live is necessary.

Zacholme7 commented:
small comments, gj otherwise!

dknopik committed:
address review and fix bugs

- `wss` -> `ws` in default EL URL
- fix slot check for validator load
- correctly decode decrypted keys
- wait for historical sync to finish before starting duties

Zacholme7 commented on anchor/eth/src/sync.rs:
```diff
@@ -106,6 +108,8 @@ pub struct SsvEventSyncer {
     /// The network the node is connected to
     network: Network,
+    /// Notify a channel as soon as the historical sync is done
+    historic_finished_notify: Option<Sender<()>>,
```
Why is this an option? I think we will always want to wait for historical sync to finish first, right?

dknopik replied:
We can only send a value through the `oneshot` sender once, consuming the sender (as it's taking ownership).

Therefore we need the option to be able to gain ownership via `take`.

Also, we might not need to wait for historic sync in e.g. testing code.

Zacholme7 commented on anchor/validator_store/src/lib.rs:
```diff
@@ -56,69 +61,173 @@ use validator_store::{
+            .collect::<HashSet<_>>();
+        let db_clusters = self.database.get_own_clusters().iter().collect::<Vec<_>>();
+
+        for (cluster, validator) in db_clusters
+            .into_iter()
+            .filter_map(|id| self.database.clusters().get_by(id.key()).map(Arc::new))
+            .flat_map(|cluster| {
+                self.database
+                    .metadata()
+                    .get_all_by(&cluster.cluster_id)
+                    .unwrap_or_default()
+                    .into_iter()
+                    .map(move |metadata| (cluster.clone(), metadata))
```
Note for future us, we should probably find some way to optimize/simplify this in the future :sweat_smile:

diegomrsantos replied:
I understand we're still learning and exploring, and I'm also doing the same, but something to keep in mind https://yiming.dev/clipping/2019/03/21/le-blanc's-law-a-k-a-later-equals-never/

dknopik replied:
I'll read that later. :P

dknopik replied:
But for real, I get what you mean. Still, I think it's worth "rushing" to an MVP, because then we can see where the problems are at, and optimize based on findings.

Zacholme7 commented:
lgtm!

dknopik committed:
Merge branch 'unstable' into wire-up

dknopik commented:
Kingy, can we please merge this? 🥺 @jking-aus

jking-aus reviewed (approved):
lgtm
